libraries/chain/controller.cpp

```diff
@@ -3391,8 +3391,10 @@ struct controller_impl {
 
    void update_peer_keys() {
       // if syncing or replaying old blocks don't bother updating peer keys
-      if (!peer_keys_db.is_active() || fc::time_point::now() - chain_head.timestamp() > fc::minutes(5))
-         return;
+      if (peer_keys_db.last_block_num_update() > 0) { // can't have been updated unless active
```

Comment:
Test seems wrong. If not active, `peer_keys_db.last_block_num_update() ` will be 0 and we won't do the `return` at line 3396, and continue to execute the try block below.

Reply:
Also the test `fc::time_point::now() - chain_head.timestamp() > fc::minutes(5))` seems wrong as well. Why would it ever be true? chain_head block more than 5 minutes old?

I think the test should be something like:

```
      if (!peer_keys_db.is_active() ||
          (peer_keys_db.last_block_num_update() > 0 && chain_head.block_num() - peer_keys_db.last_block_num_update() > 600)
            return;
      }
```

Reply:
Good point about checking `is_active()`.

> Also the test `fc::time_point::now() - chain_head.timestamp() > fc::minutes(5))` seems wrong as well. Why would it ever be true? chain_head block more than 5 minutes old?

It is true when syncing or replaying blocks as indicated by the comment.

Reply:
So why the `peer_keys_db.last_block_num_update() > 0`? Do you want to do the update once at the beginning even if we are replaying?

Reply:
If we are replaying (and we don't know how far behind we are), we have an old state in chainbase, so why get old IP from chainbase and gossip them? Especially since we get the oldest peer IPs at startup, and then don't update them until we're fully caught up.

I think we should wait till we're caught up to start updating peer keys.

Reply:
I was concerned about bootstrapping a dead/paused chain. But in that situation manual BP connections can be used and would likely be used anyway.

Reply:
https://github.com/AntelopeIO/spring/pull/1503/commits/2c965d98b726d3da588dc728fb4317059c0ceef1